CommonLib/Error/GREYErrorFormatter.m

```diff
@@ -0,0 +1,141 @@
+//
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "GREYErrorFormatter.h"
+
+#import "GREYObjectFormatter.h"
+#import "GREYInteraction.h"
+#import "GREYErrorConstants.h"
+#import "NSError+GREYCommon.h"
+#import "GREYError+Private.h"
+#import "GREYFatalAsserts.h"
+
+#pragma mark - UI Hierarchy Keys
+
+static NSString *const kHierarchyWindowLegendKey                 = @"[Window 1]";
+static NSString *const kHierarchyAcessibilityLegendKey           = @"[AX]";
+static NSString *const kHierarchyUserInteractionEnabledLegendKey = @"[UIE]";
+static NSString *const kHierarchyBackWindowKey                   = @"Back-Most Window";
+static NSString *const kHierarchyAccessibilityKey                = @"Accessibility";
+static NSString *const kHierarchyUserInteractionEnabledKey       = @"User Interaction Enabled";
+static NSString *const kHierarchyLegendKey                       = @"Legend";
+static NSString *const kHierarchyHeaderKey                       = @"UI Hierarchy (ordered by wind"
+                                                                   @"ow level, back to front):\n";
+
+#pragma mark - GREYErrorFormatter
+
+@implementation GREYErrorFormatter
+
+#pragma mark - Public Methods
+
++ (NSString *)formattedDescriptionForError:(GREYError *)error {
+  if (GREYShouldUseErrorFormatterForError(error)) {
+    return LoggerDescription(error);
+  }
+  return [GREYObjectFormatter formatDictionary:[error grey_descriptionDictionary]
+                                        indent:kGREYObjectFormatIndent
+                                     hideEmpty:YES
+                                      keyOrder:nil];
+}
+
+#pragma mark - Public Functions
+
+BOOL GREYShouldUseErrorFormatterForError(GREYError *error) {
+  return [error.domain isEqualToString:kGREYInteractionErrorDomain] &&
+          error.code == kGREYInteractionElementNotFoundErrorCode;
+}
+
+BOOL GREYShouldUseErrorFormatterForExceptionReason(NSString *reason) {
```

Comment:
Looking at https://github.com/willsaid-google/EarlGrey/pull/6/files, this function will definitely become bloated. Instead, we could add a prefix to this formatter hierarchy as we roll it out piecemeal per error.

So, in LoggerDescription - you can start the string with:

"GREYErrorFormatter Information:\n" and trim the prefix in the failure handler as a temporary hack. It will remove the need to add more strings in here.

Reply:
Trimming the prefix isn't reliable since, for example, a user could call GREYError's description. At what point would we know when to trim?...Only when the default failure handler is used. So, i propose we use a flag that is actually useful, and won't need to be trimmed, such as "EarlGrey Encountered an Error:"

Reply:
Pushed. Please check out the changes